### PR TITLE
feat(storage): Improve optimize table compact

### LINKED ISSUE #6372: Improvement: improve optimize table compact
improve optimize table compact

### PR DESCRIPTION
**Pull request by zhyass**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

The previous compact strategy was to insert the data overwrite directly, it's a "simple and violent" strategy.

Now it is modified to select the appropriate block for compact, and then reorganize the blocks into new segments.

## Changelog

- Improvement
- Not for changelog (changelog entry is not required)

## Related Issues

Fixes #6372

### CODE REVIEW

**Comment by BohuTANG** [comment]
Hi @zhyass 
Please add more summaries about this PR to make it more clear to your reviewer.
Thanks

File: query/src/storages/fuse/operations/mutation/compact_mutator.rs [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+                }
+            }
+        }
+        let remains = compactor.finish()?;
+        if let Some(blocks) = remains {
+            for block in blocks {
+                let new_block_meta = block_writer.write(block).await?;
+                remain_blocks.push(new_block_meta);
+            }
+        }
```
Comment (dantengsky) [inline query/src/storages/fuse/operations/mutation/compact_mutator.rs:141]: duplicated code? (line 128 ~ 132)

File: query/src/storages/fuse/operations/compact.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                    cache.put(loc, Arc::new(new_snapshot));
+                }
+                Ok(())
+            }
+            Err(e) => {
+                // commit snapshot to meta server failed, try to delete it.
+                // "major GC" will collect this, if deletion failure (even after DAL retried)
+                let _ = operator.object(&loc).delete().await;
+                Err(e)
+            }
```
Comment (dantengsky) [inline query/src/storages/fuse/operations/compact.rs:85]: seems to be duplicated with codes in operations/commit.rs?

File: query/src/storages/fuse/operations/mutation/compact_mutator.rs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        // write the new segment out (and keep it in undo log)
+        let snapshot_loc = self.location_generator.snapshot_location_from_uuid(
+            &new_snapshot.snapshot_id,
+            new_snapshot.format_version(),
+        )?;
+        let bytes = serde_json::to_vec(&new_snapshot)?;
+        self.data_accessor
+            .object(&snapshot_loc)
+            .write(bytes)
+            .await?;
```
Comment (dantengsky) [inline query/src/storages/fuse/operations/mutation/compact_mutator.rs:84]: `object(..).write(..)` may failed for some transient io errors,  maybe we should use `io::write_meta` instead?

https://github.com/datafuselabs/databend/blob/main/query/src/storages/fuse/operations/mutation/mutations_collector.rs#L138-L142

File: query/src/storages/fuse/operations/mutation/compact_mutator.rs [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+use crate::storages::fuse::io::TableMetaLocationGenerator;
+use crate::storages::fuse::meta::Location;
+use crate::storages::fuse::meta::SegmentInfo;
+use crate::storages::fuse::meta::Statistics;
+use crate::storages::fuse::meta::TableSnapshot;
+use crate::storages::fuse::statistics::reducers::reduce_block_metas;
+use crate::storages::fuse::statistics::reducers::reduce_statistics;
+use crate::storages::fuse::FuseTable;
+
+pub struct CompactMutator<'a> {
```
Comment (dantengsky) [inline query/src/storages/fuse/operations/mutation/compact_mutator.rs:35]: Naming is much better than `DeletionCollector` :+1:

**Comment by dantengsky** [comment]
@zhyass :+1:  finally, the `naive` impl of compaction is being replaced.  left some comments, please have a look

**Pull request merged**